src/snowflake/cli/api/utils/definition_rendering.py

```diff
@@ -0,0 +1,163 @@
+from __future__ import annotations
+
+import os
+from collections import deque
+from dataclasses import dataclass
+from typing import Optional
+
+import yaml
+from jinja2 import Environment, UndefinedError, nodes
+from snowflake.cli.api.utils.graph import Graph
+from snowflake.cli.api.utils.rendering import get_snowflake_cli_jinja_env
+
+
+class Variable:
+    def __init__(self, vars_chain):
+        self._vars_chain = list(vars_chain)
+        self.value: str | None = None
+
+    def get_vars_hierarchy(self) -> deque[str]:
+        return deque(self._vars_chain)
+
+    def get_key(self):
+        return ".".join(self._vars_chain)
+
+    def __hash__(self):
+        return hash(self.get_key())
+
+    def __eq__(self, other):
+        return self.get_key() == other.get_key()
+
+
+@dataclass(eq=False)
+class TemplateGraphNode(Graph.Node):
+    templated_value: str | None = None
+    rendered_value: str | None = None
+    variable: Variable | None = None
+
+    def store_in_context(self, context):
+        # if graph node is ctx.env.test with value x, context will be built as {'ctx': {'env': {'test': 'x'}}}
+        vars_chain = self.variable.get_vars_hierarchy()
+        self._store_in_context_recursive(context, vars_chain)
+
+    def _store_in_context_recursive(self, context, vars_chain: deque):
+        if len(vars_chain) == 0:
+            return self.rendered_value
+
+        current_level_key = vars_chain.popleft()
+        current_value_in_key = (
+            context[current_level_key] if current_level_key in context else {}
+        )
+        context[current_level_key] = self._store_in_context_recursive(
+            current_value_in_key, vars_chain
+        )
+        return context
+
+
+def _get_referenced_vars(
+    ast_node, variable_attr_chain: deque = deque()
+) -> set[Variable]:
+    # Var nodes end in Name node, and starts with Getattr node.
+    # Example: ctx.env.test will look like Getattr(test) -> Getattr(env) -> Name(ctx)
+    all_referenced_vars = set()
+
+    variable_appended = False
+    if isinstance(ast_node, nodes.Getattr):
+        variable_attr_chain.appendleft(getattr(ast_node, "attr"))
+        variable_appended = True
+    elif isinstance(ast_node, nodes.Name):
+        variable_attr_chain.appendleft(getattr(ast_node, "name"))
```

Comment:
why not just ast_node.name? The `getattr` call is usually necessary only when you don't know the name a priori, but here we're using it with a constant.

Reply:
mypy was complaining about name not being defined, so I was forced to use that.

Reply:
Would something like `# type: ignore` help then (or type hints)? In cases where the tooling fails us badly, then we tend to prefer disabling the tooling over crippling the code.